ast/check.go

```diff
@@ -652,72 +657,50 @@ func (rc *refChecker) checkRef(curr *TypeEnv, node *typeTreeNode, ref Ref, idx i
 
 	head := ref[idx]
 
-	// Handle constant ref operands, i.e., strings or the ref head.
-	if _, ok := head.Value.(String); ok || idx == 0 {
-
-		child := node.Child(head.Value)
-		if child == nil {
-
-			if curr.next != nil {
-				next := curr.next
-				return rc.checkRef(next, next.tree, ref, 0)
-			}
-
-			if RootDocumentNames.Contains(ref[0]) {
-				return rc.checkRefLeaf(types.A, ref, 1)
-			}
-
-			return rc.checkRefLeaf(types.A, ref, 0)
+	// NOTE(sr): as long as package statements are required, this isn't possible:
+	// the shortest possible rule ref is data.a.b (b is idx 2), idx 1 and 2 need to
+	// be strings or vars.
+	if idx == 1 || idx == 2 {
+		switch head.Value.(type) {
+		case Var, String: // OK
+		default:
+			have := rc.env.Get(head.Value)
+			return newRefErrInvalid(ref[0].Location, rc.varRewriter(ref), idx, have, types.S, getOneOfForNode(node))
 		}
-
-		if child.Leaf() {
-			return rc.checkRefLeaf(child.Value(), ref, idx+1)
-		}
-
-		return rc.checkRef(curr, child, ref, idx+1)
 	}
 
-	// Handle dynamic ref operands.
-	switch value := head.Value.(type) {
-
-	case Var:
-
-		if exist := rc.env.Get(value); exist != nil {
-			if !unifies(types.S, exist) {
-				return newRefErrInvalid(ref[0].Location, rc.varRewriter(ref), idx, exist, types.S, getOneOfForNode(node))
+	if v, ok := head.Value.(Var); ok && idx != 0 {
+		tpe := types.Keys(rc.env.getRefRecExtent(node))
+		if exist := rc.env.Get(v); exist != nil {
+			if !unifies(tpe, exist) {
+				return newRefErrInvalid(ref[0].Location, rc.varRewriter(ref), idx, exist, tpe, getOneOfForNode(node))
 			}
 		} else {
-			rc.env.tree.PutOne(value, types.S)
+			rc.env.tree.PutOne(v, tpe)
 		}
+	}
 
-	case Ref:
+	child := node.Child(head.Value)
+	if child == nil {
+		// NOTE(sr): idx is reset on purpose: we start over
+		switch {
+		case curr.next != nil:
+			next := curr.next
+			return rc.checkRef(next, next.tree, ref, 0)
 
-		exist := rc.env.Get(value)
-		if exist == nil {
-			// If ref type is unknown, an error will already be reported so
-			// stop here.
-			return nil
-		}
+		case RootDocumentNames.Contains(ref[0]):
+			return rc.checkRefLeaf(types.A, ref, 1)
 
-		if !unifies(types.S, exist) {
-			return newRefErrInvalid(ref[0].Location, rc.varRewriter(ref), idx, exist, types.S, getOneOfForNode(node))
+		default:
+			return rc.checkRefLeaf(types.A, ref, 0)
 		}
-
-	// Catch other ref operand types here. Non-leaf nodes must be referred to
-	// with string values.
-	default:
-		return newRefErrInvalid(ref[0].Location, rc.varRewriter(ref), idx, nil, types.S, getOneOfForNode(node))
 	}
 
-	// Run checking on remaining portion of the ref. Note, since the ref
-	// potentially refers to data for which no type information exists,
-	// checking should never fail.
-	node.Children().Iter(func(_, child util.T) bool {
-		_ = rc.checkRef(curr, child.(*typeTreeNode), ref, idx+1) // ignore error
-		return false
-	})
```

Comment:
So I've reworked this portion, and this ☝️ `node.Children().Iter(...)` call got removed. I didn't know where to put it 😅 and it didn't seem to have any effect either.

From what I understand, it would let us learn things about variable's types if they're used in refs under some condition: (Im describing the previous code to myself)

1. the value isn't a string, and it's not the first one (idx=0) -- that's jumping past the first switch block
2. the value is a var that we've either seen before and that "checks out", or we now learnt that it's a string
3. we're not at the end of our ref yet (so idx < len(ref))

So....the type tree node has children, but we haven't found one for value, since it's a var.

With the current new code, that combination of conditions will get us into `rc.checkRefLeaf(types.A, ref, 0)`, I think... which should have the same effect 🤔 I might be wrong though.

@tsandall What do you think?

Reply:
I wonder if the `node.Children().Iter(...)`, or at least its index-advancing behavior might be significant? :thinking:

Reply:
It's been a while but I think the rationale was that if:

* The type checker has not reached a leaf node, and...
* The current ref element is a variable, then...
* The expression potentially refers to a base document.

Since the expression potentially refers to a base document, any uninferred types will get assigned ANY and any existing types will unify.

If the new implementation deals with that case, then we're good (AFAICT).

Reply:
OK thanks a lot for that -- found a test case that passes on main but fails here, in the discussed scenario: https://github.com/open-policy-agent/opa/compare/main...srenatus:opa:sr/ast/check/add-test-case

I'll adapt this. 🛠️